### PR TITLE
Add Objectives and Configuration for PickNik UR Robots

### PR DESCRIPTION
**Pull request by noah-wardlow**
[sc-13924]

this is slated for 2.2, with an additional PR to the Studio repo that will throw toasts on select point objectives if there are no cameras

### CODE REVIEW

**Review by sea-bass — APPROVED** [review]

File: src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py [hunk 1]
```diff
@@ -33,23 +33,37 @@
 
 from ament_index_python import get_package_share_directory
```
Comment (sea-bass) [inline src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py:34]: I might push back on the changes to this file. When moving this to this repo, I took care to remove certain things that may not have been needed, so I'll fill in more details in more comments.

File: src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py [hunk 2]
```diff
@@ -59,7 +73,15 @@ def path_pattern_change_for_ignition(urdf_string):
 
 
 def generate_simulation_description(context, *args, **settings):
-    world_name = settings.get("gazebo_world_name", "space_station.sdf")
+    nodes = []
+    is_test = LaunchConfiguration("is_test").perform(context).lower() == "true"
```
Comment (sea-bass) [inline src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py:77]: If we're not running this configuration through our CI pipeline, we can get rid of the `is_test` argument as well as the toggle on world name key below.

File: src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py [hunk 3]
```diff
@@ -59,7 +73,15 @@ def path_pattern_change_for_ignition(urdf_string):
 
 
 def generate_simulation_description(context, *args, **settings):
-    world_name = settings.get("gazebo_world_name", "space_station.sdf")
```
Comment (sea-bass) [inline src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py:62]: If we don't want to switch to other world files as a launch argument (which probably isn't necessary at this point), we can get rid of these options.

File: src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py [hunk 4]
```diff
@@ -81,31 +103,57 @@ def generate_simulation_description(context, *args, **settings):
...
 
+        # If no display is available, set up a Xvfb for headless rendering
+        if not os.environ.get("DISPLAY"):
+            print("Adding Xvfb for simulation...")
+            display_id = ":99"
+            os.environ["DISPLAY"] = display_id
+            xvfb = ExecuteProcess(
+                name="xvfb", cmd=["Xvfb", display_id, "-screen", "0", "1600x1200x16"]
+            )
+            nodes.append(xvfb)
```
Comment (sea-bass) [inline src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py:122]: Again, if not running as part of CI we can remove

File: src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py [hunk 5]
```diff
@@ -81,31 +103,57 @@ def generate_simulation_description(context, *args, **settings):
...
+
     robot_urdf_ignition = path_pattern_change_for_ignition(robot_urdf)
 
-    # Launch Gazebo
+    # Include URDF
+    scene_xacro_path = get_ros_path(
+        "picknik_ur_gazebo_config", "description/simulation_scene.urdf.xacro"
+    )
+    scene_urdf = xacro_to_urdf(scene_xacro_path, None)
+    scene_urdf_ignition = path_pattern_change_for_ignition(scene_urdf)
```
Comment (sea-bass) [inline src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py:154]: this code is duplicated in that I had moved it below, near where it's use for launching. You can undo this part, or remove the other below. Either/or

File: src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py [hunk 6]
```diff
@@ -317,4 +361,4 @@ def generate_launch_description():
             spawn_scene,
             camera_transforms_node,
         ]
-    )
+    )
```
Comment (sea-bass) [inline src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py:364]: missing trailing newline

**Review by sea-bass — COMMENTED** [review]
Tested in mock hardware and Gazebo, and those works great!

Regular config worked totally fine for me.

In the Gazebo config, the `Pick` waypoint is still a bit too low to the table. It doesn't collide, but then Pick Object doesn't find cuboids because the camera is too close to the table. I zoomed out a bit and all worked fine.

Also I had simplified the Gazebo launch file by removing all the things that were only used for CI (which I believe we don't do anymore in favor of mock hardware), so I would revert those changes.

Lastly: Should we also be testing these configs on hardware, along with the other PR, and if so, some detailed instructions would be good.

File: src/picknik_ur_base_config/objectives/looping_pick_and_place_object.xml [hunk 7]
```diff
@@ -0,0 +1,95 @@
+<?xml version="1.0"?>
+<root BTCPP_format="4" main_tree_to_execute="Looping Pick and Place Object">
+    <!-- ////////// -->
+    <BehaviorTree ID="Looping Pick and Place Object">
```
Comment (MikeWrock) [inline src/picknik_ur_base_config/objectives/looping_pick_and_place_object.xml:4]:
```suggestion
    <BehaviorTree ID="Looping Pick and Place Object" _description="Repeatedly attempt to detect and grasp a cuboid, then place it down">
```
Perhaps you have a better description, but this objective is the only one without

File: src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py (hunk 1 above)
Comment (eholum) [reply src/picknik_ur_gazebo_config/launch/sim/hardware_sim.launch.py]: Understood!

I reverted all but the critical and minor functional changes.

File: src/picknik_ur_base_config/description/picknik_ur5e.xacro [hunk 8]
```diff
@@ -107,7 +112,7 @@
     <origin rpy="0 0 0" xyz="0 0 0" />
   </joint>
 
-  <!-- External Camera -->
+  <!-- Camera External -->
```
[the remark this answers is not in the document]
Comment (eholum) [reply src/picknik_ur_base_config/description/picknik_ur5e.xacro]: Done!

**Comment by eholum** [comment]
> In the Gazebo config, the Pick waypoint is still a bit too low to the table. It doesn't collide, but then Pick Object doesn't find cuboids because the camera is too close to the table. I zoomed out a bit and all worked fine.

Cool. I pushed it back even more, and adjusted the GAC params so that I can actually (sometimes) get the cuboid to stick in the gripper. Let me know if that's better for you?

> Also I had simplified the Gazebo launch file by removing all the things that were only used for CI.

I reverted everything except for the params we identified as needing to change.

**Review by sea-bass — APPROVED** [review]
This LGTM!

File: src/picknik_ur_base_config/objectives/get_door_handle_config.yaml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+GetDoorHandle:
```
Comment (chancecardona) [inline src/picknik_ur_base_config/objectives/get_door_handle_config.yaml:1]: PR https://github.com/PickNikRobotics/moveit_studio/pull/2974 looks to be partially merged in here, making this file a duplicate / unused. Could we remove this file and the `setup_mtc_open_lever_handle_door_config.yaml` files entirely?

File: src/picknik_ur_base_config/objectives/setup_mtc_open_lever_handle_door_config.yaml [hunk 10]
```diff
@@ -0,0 +1,23 @@
+SetupMTCOpenLeverHandleDoor:
```
Comment (chancecardona) [inline src/picknik_ur_base_config/objectives/setup_mtc_open_lever_handle_door_config.yaml:1]: Could we remove this file as well? No longer needed.

**Review by chancecardona — CHANGES_REQUESTED** [review]
Overall this looks pretty good. The `Pick` waypoint is still a little bit off, I get "Behavior: PlanMTCTask. Error: Failed to execute action: Objective planning failed." in both mock and Gz. The other waypoints look good to me.

Everything else has been running well for me in mock and Gazebo so far. A few comments on unused config files.

File: src/picknik_ur_base_config/objectives/get_door_handle_config.yaml (hunk 9 above)
Comment (eholum) [reply src/picknik_ur_base_config/objectives/get_door_handle_config.yaml]: Good catch! Removed.

File: src/picknik_ur_base_config/objectives/setup_mtc_open_lever_handle_door_config.yaml (hunk 10 above)
Comment (eholum) [reply src/picknik_ur_base_config/objectives/setup_mtc_open_lever_handle_door_config.yaml]: Also removed.

**Comment by eholum** [comment]
> The `Pick` waypoint is still a little bit off, I get "Behavior: PlanMTCTask. Error: Failed to execute action: Objective planning failed."

Are you sure? Have you reset the waypoints between deployments? There is not a "Pick" waypoint for the ur site config.

**Comment by eholum** [comment]
> Pick in Gazebo works fine now when I'm testing, but it is a good deal above the actual block. Same with Grab on mock for the site config. 

Yes, the intent here is to pose the arm in a reasonable pose to execute the pick objective, rather than actually contact the block in the GZ config. I think we're good to move forward with that.

**Review by chancecardona — DISMISSED** [review]
LGTM!

**Review by sea-bass — DISMISSED** [review]

**Review by eholum — DISMISSED** [review]
Approving with the 2.1.1 release. @chancecardona @sea-bass is this still good to go?

**Review by eholum — APPROVED** [review]
Re-approving.

**Review by sea-bass — APPROVED** [review]
still good to go!

**Pull request merged**